src/garn-bin.js

```diff
@@ -1,86 +1,257 @@
-    rootPath = path.join(basePath, '..', '..');
+  const argv = minimist(process.argv.slice(2));
+
+  const buildsystemPath = argv[buildsystemPathArgName];
+  const buildCache = '.buildcache';
+  const buildCachePath = path.join(buildsystemPath, buildCache);
+  const buildCacheManifestPath = path.join(buildCachePath, '.manifest.json');
+  const basePath = path.join(buildsystemPath, '..');
+  let rootPath = basePath;
+  if (!fs.existsSync(path.join(basePath, 'packages', 'garn.cmd'))) {
+    const parentName = path.basename(path.join(basePath, '..'));
+    if (parentName === 'packages') {
+      rootPath = path.join(basePath, '..', '..');
+    }
   }
-}
 
-const needsCompile = anyFileInManifestHasChanged();
-let writeMetaData = false;
+  const yarnLockPath = path.join(rootPath, 'yarn.lock');
+  const packageLockJsonPath = path.join(rootPath, 'package-lock.json');
+  const copiedYarnLockPath = path.join(buildCachePath, '.yarn.lock');
+  const copiedPackageLockJsonPath = path.join(buildCachePath, '.package-lock.json');
 
-const skipCompile = 'asap' in argv && fs.existsSync(buildCacheManifestPath);
+  const shouldRestore = shouldRestoreNpmPackages(
+    packageLockJsonPath,
+    yarnLockPath,
+    copiedPackageLockJsonPath,
+    copiedYarnLockPath,
+  );
+  if (shouldRestore === false) {
+    compileIfNeededAndRun(argv, rootPath, buildsystemPath, buildCache, buildCachePath, buildCacheManifestPath);
+  } else {
+    restoreNpmPackages(
+      rootPath,
+      shouldRestore,
+      packageLockJsonPath,
+      yarnLockPath,
+      copiedPackageLockJsonPath,
+      copiedYarnLockPath,
+    );
+  }
+}
 
-if ((!skipCompile && needsCompile) || 'compile-buildsystem' in argv) {
-  writeMetaData = true;
-  console.log('Compiling buildsystem...');
-  compile();
+/**
+ * @param {string} rootPath
+ * @param {'yarn' | 'npm'} restorePackagesWith
+ * @param {string} packageLockJsonPath
+ * @param {string} yarnLockPath
+ * @param {string} copiedPackageLockJsonPath
+ * @param {string} copiedYarnLockPath
+ */
+function restoreNpmPackages(
+  rootPath,
+  restorePackagesWith,
+  packageLockJsonPath,
+  yarnLockPath,
+  copiedPackageLockJsonPath,
+  copiedYarnLockPath,
+) {
+  const restartGarn = () => {
+    const garnPath = path.join(rootPath, 'node_modules', '.bin', 'garn' + execExt);
+    const garn = childProcess.spawn(garnPath, ['--force'], {
+      cwd: process.cwd(),
+      stdio: 'inherit',
+    });
+    garn.on('exit', exitCode => process.exit(exitCode));
+  };
 
-  // Temp code to copy package.json into .buildcache/
-  const allPackagesPath = path.join(rootPath, 'packages');
-  let packagesPath = path.join(buildCachePath, 'packages');
-  if (!fs.existsSync(packagesPath)) {
-    packagesPath = path.join(packagesPath, '..');
+  if (restorePackagesWith === 'yarn') {
+    let yarnPath = path.join(rootPath, 'yarn' + execExt);
+    if (!fs.existsSync(yarnPath)) {
+      yarnPath = 'yarn' + execExt;
+    }
+    const yarn = childProcess.spawn(yarnPath, ['--force'], {
```

Comment:
Is `--force` necessary here? That flag isn't supported in yarn 3 and it throws an error when you use it.